### ISSUE TITLE
Incorrect result of argmax on CUDA with large non-contig tensors and on non-contig dimensions

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#32863]
```python
import torch

a = torch.load('bug.pt', 'cuda') # giant uint8 tensor with binary 0/1 values
C = 11359
b = a[:, 0, C]

print('shape', b.shape)
print('argmax', b.argmax())
print('max', b.max())

K = int(b.argmax())

print(b[K]) # should be 1, but is 0
print(a[K, 0, C]) # should be 1, but is 0

print(b.nonzero()) # different result compared to argmax
print('bad argmax', a.argmax(dim = 0)[0][C])
```
```
>>> torch.__version__
'1.4.0'
>>> torch.version.git_version
'143868c3df4eb2acdcb166b340a1063abf61339c'
```

I tried to minimize the size of `bug.pt`, but my first attempts led to bug disappearing, so I'm uploading it to my OneDrive as is (it's quite heavy - 3Gb): https://1drv.ms/u/s!Apx8USiTtrYmqfhMddRNKxibLatubA?e=wyihRU 

(it would be nice for such situations to have a safe unpickler that can't execute arbitrary code or have it support loading/saving tensors from/to hdf5 as well. doesn't pytorch have some rudimentary unpickler for TorchCpp use?)

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by ezyang** [comment]
High priority due to silently incorrect results

**Comment by vadimkantorov** [comment]
@ezyang I wonder if related functions like argmin/argsort/topk/max/min/sort are also implicated

**Comment by ngimel** [comment]
argmin almost certainly, the rest unlikely because they take different code paths

**Comment by vadimkantorov** [comment]
Even argmin? Anyway, if we figure out the offending tensor layout / size, it should be easy to add a test for several of them

**Comment by ngimel** [comment]
self-contained repro
```
import torch
  
C = 11359
bc = torch.zeros(74331,dtype=torch.uint8, device="cuda")
bc[38000]=1
a=torch.zeros(74331,1,44036, dtype = torch.uint8, device="cuda")
asliced = a[:, :, ::2]
asliced[:,0,C] = bc
bb=asliced[:,0,C]
print('shape', bb.shape)
print('argmax', bb.argmax(), bc.argmax())
print('max', bb.max(), bc.max())

K = int(bb.argmax())
Kc = int(bc.argmax())
print(bb[K],  bc[Kc]) # should be 1, but is 0
print(asliced[K, 0, C], asliced[Kc, 0, C]) # should be 1, but is 0

print(bb.nonzero()) # different result compared to argmax
print('bad argmax', asliced.argmax(dim = 0)[0][C])
```

**Comment by vadimkantorov** [comment]
When I added contiguous/clone call in b definition, the issue disappeared, iirc. Can you check this too? If true, contiguity still matters

**Comment by ngimel** [comment]
Yep, that's true, as you see in my repro there's a contiguous tensor `bc` that produces the right result, and a discontiguous slice `bb`, where results are wrong. The discontiguity in original "full-size" tensor `a` does not matter though.

**Comment by vadimkantorov** [comment]
Right. Even In my case the original tensor was contiguous iirc, only the slice was discontiguous

**Comment by vadimkantorov** [comment]
Or maybe stride is considered to be 1 somewhere, since plain argmax over dim 0 is also failing...

**Comment by ngimel** [comment]
The root cause is TensorIterator splitting tensors that cannot be indexed with int32, and processing them chunk by chunk. For everything except reductions involving indices this works good, however, for reductions involving indices, such as argmax or argmin, the indexing in the split tensor starts with 0 rather than with whatever index there was in the original large tensor. In the example below:
```.py
import torch

bc = torch.zeros(2000,dtype=torch.uint8, device="cuda")
bc[1355]=1
a=torch.zeros(2000,1,45000*30, dtype = torch.uint8, device="cuda")
a[:,0,0] = bc
bb=a[:,0,0]
print('argmax',bb.argmax()) #prints 355
torch.cuda.synchronize()
```
tensor is split into 2 equal parts with size 1000 along 0-th dimension, the max element is the 355th element in the second part, and its index is recorded as 355 instead of 1355. I'm not sure at this point how to fix.

**Comment by vadimkantorov** [comment]
Does it finally also implicate max/min with indices?

**Comment by ngimel** [comment]
No, min and max with indices don't go through tensorIterator, they use legacy codepaths in THC iirc.

**Comment by vadimkantorov** [comment]
@ngimel If only argmax uses TensorIterator with indices, a temporary fix could be calling reusing max() code path for CUDA as well (like it seems already done on CPU)

**Comment by ezyang** [comment]
ngimel thinks there are not good fixes. TensorIterator is being used for reductions, and it splits the tensors into smaller tensors. But for argmax/argmin, you're interested in the value of the index, which when split doesn't make sense. So we'd have to bring back the path for reduction and index it with int64... but that's a perf hit. ezyang wonders if you couldn't maintain the base offset of the index (int64), but ngimel suspects this will slow things down because index is passed everywhere in TensorIterator. Also, TensorIterator splits the tensor recursively, and there's no guarantee that it's split on the same dimension so it's not well defined what the offset is.

**Comment by vadimkantorov** [comment]
Should argmax impl be replaced by the `max(dim = dim).indices` code path in the meanwhile?

**Comment by ezyang** [comment]
Seems like an easy enough fix to do for now.

**Comment by vadimkantorov** [comment]
argmin as well...

**Comment by peterbell10** [comment]
I tried the suggested workaround but it seems that `max` is also broken when indexing further than 2^32.
```python
>>> x = torch.zeros(2**32, device='cuda')
>>> x[-1] = 1.
>>> x.max(0)
torch.return_types.max(
values=tensor(0., device='cuda:0'),
indices=tensor(0, device='cuda:0'))
```
`max` and `min` are implemented by `redudeDimIndex` in `THCTensorMathReduce.cuh` which silently casts the indexes to 32-bit `unsigned`:
https://github.com/pytorch/pytorch/blob/de27f4261d3742c124f4a9721ae98c3e87f6c640/aten/src/THC/THCTensorMathReduce.cuh#L490

**Comment by peterbell10** [comment]
We could fix `THC` by either using 64-bit indexing everywhere or by templating `kernelTransformReduceInntermostDimIndex` and `kernelTransformReduceOuterDimIndex` on the index type. This would allow the suggested workaround for `argmin/max`.

To fix `TensorIterator` I see that the reduction already has a flag `accumulate` that triggers additional `combine` steps when reducing along a split axis. Perhaps we could add a separate operation to `ops_t` that covers transformation from local to global indexes. For most ops this would be the identity and should be optimised out, for `argmax/min` it would be something like:
```
arg_t translate_index(arg_t arg, int64_t base_idx) {
  return {a.first, a.second + base_idx};
}
```

**Comment by vadimkantorov** [comment]
> I tried the suggested workaround but it seems that `max` is also broken when indexing further than 2^32.

Could you please also test topk(...).indices for your torch.max case and argsort for the case of slice (where the indices are in int64, but the number of elements across the given dim is not so large - like in my original example)?

Maybe they are also broken because of the THC index problem or another similar indexing bug

**Comment by peterbell10** [comment]
- `topk` and `argsort` both worked fine, looking at the code they have dedicated code paths for 64-bit indexing.
- `max` worked for @ngimel's example. `THC_reduceDimIndex` calls `THCTensor_newContiguous` on the `src` tensor, so it isn't indexing into the larger tensor at all. Still broken when operating on the full tensor though.

**Comment by vadimkantorov** [comment]
The last question is whether sort(dim) does 64bit indexing in @ngimel's example :) If argsort uses sort under the hood, then it should be fine I guess

**Comment by peterbell10** [comment]
Yes, `argsort` is implemented in terms of `sort`.

**Comment by vadimkantorov** [comment]
@peterbell10 Is there a similar bug in TH on CPU?

**Comment by peterbell10** [comment]
I haven't checked all of the CPU functions but `max` and `min` at least work.  I suspect all of TH is fine since 32-bit indexing doesn't have any performance benefit.
